WordPress/Classes/ViewRelated/Tools/SettingsTextViewController.m

```diff
@@ -24,37 +32,130 @@ - (void)dealloc
 {
     self = [super initWithStyle:UITableViewStyleGrouped];
     if (self) {
         _text = text;
         _placeholder = placeholder;
         _hint = hint;
-        _isPassword = isPassword;
     }
     return self;
 }
 
-- (instancetype)initWithStyle:(UITableViewStyle)style
+
+#pragma mark - View Lifecycle
+
+- (void)viewDidLoad
+{
+    [super viewDidLoad];
+    [WPStyleGuide resetReadableMarginsForTableView:self.tableView];
+    [WPStyleGuide configureColorsForView:self.view andTableView:self.tableView];
+}
+
+- (void)viewWillAppear:(BOOL)animated
 {
-    return [self initWithText:@"" placeholder:@"" hint:@"" isPassword:NO];
+    [super viewWillAppear:animated];
+    [self setupNavigationButtonsIfNeeded];
+    [self startListeningToNotifications];
 }
 
 - (void)viewDidAppear:(BOOL)animated
 {
-    [self.textField becomeFirstResponder];
     [super viewDidAppear:animated];
+    [self.textField becomeFirstResponder];
 }
 
-- (void)viewDidLoad
+- (void)viewWillDisappear:(BOOL)animated
 {
-    [super viewDidLoad];
-    [WPStyleGuide resetReadableMarginsForTableView:self.tableView];
-    [WPStyleGuide configureColorsForView:self.view andTableView:self.tableView];
+    [super viewWillDisappear:animated];
+    [self.view endEditing:YES];
+    
+    [self stopListeningToNotifications];
+    [self notifyValueDidChangeIfNeeded];
+}
+
+
+#pragma mark - NavigationItem Buttons
+
+- (void)setupNavigationButtonsIfNeeded
+{
+    if (self.shouldDisplayNavigationButtons == NO) {
+        return;
+    }
+    
+    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
+                                                                                          target:self
+                                                                                          action:@selector(cancelButtonWasPressed:)];
+    
+    
+    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone
+                                                                                           target:self
+                                                                                           action:@selector(doneButtonWasPressed:)];
+}
+
+- (BOOL)shouldDisplayNavigationButtons
+{
+    return self.isModal;
+}
+
+- (IBAction)cancelButtonWasPressed:(id)sender
+{
+    // Make sure the callback doesn't get hit
+    self.onValueChanged = nil;
+    [self dismissViewController];
+}
+
+- (IBAction)doneButtonWasPressed:(id)sender
+{
+    [self notifyValueDidChangeIfNeeded];
+    [self dismissViewController];
+}
+
+
+#pragma mark - Notifications
+
+- (void)startListeningToNotifications
+{
+    if (self.shouldDisplayNavigationButtons == false) {
+        return;
+    }
+    
+    NSNotificationCenter *nc = [NSNotificationCenter defaultCenter];
+    [nc addObserver:self selector:@selector(validateTextInput:) name:UITextFieldTextDidChangeNotification object:_textField];
```

Comment:
TIL about `UITextFieldTextDidChangeNotification` :wink: . You could also use `addTarget` which doesn't need unregistering:

```
[_textField addTarget:self 
              action:@selector(validateTextInput:) 
    forControlEvents:UIControlEventEditingChanged];
```